### PR TITLE
Adds a research log and the ability for the RD to disable point generation, makes the R&D servers stop working if they lose power.

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
* Overhauls the gui for the Research Sever controller (see below), and adds a bit more info about each server. Allows the RD, or anyone with related access, to shut off servers and thus disable point generation (emag to bypass access restriction). Adds temperature readout for each server, for fluff and possibly future relevance. Servers are listed as "Offline" if they are under EMP effects or have no power. Servers now have a random four-digit hex number following their name, to help tell them apart.

* Adds a full list of all research history in a table below the server readout on the server controller, sorted with the newest on top. Lists the name of tech, cost, name of the researcher, and the location they researched from.

* Adds in code to mark servers without power as not working. This seems like an unnoticed bug, to be honest.

* Animates the working server sprite, and adds a disabled and powered-off sprite.

The "research_log" list variable for /datum/techweb is commented as being meant for IC use, and was being updated already, but as far as I can tell was never used anywhere. So I have re-organized it a bit to feed the prior-mentioned table of research history.

## Why It's Good For The Game
Gives the RD a light bit more control over research, allows them to cut off passive point generation if people piss them off, and gives purpose to the server controller console.

## Changelog
:cl:
add: The R&D Server Controller console now lists the entire research history, including names of people who researched each item and locations it was done from.
add: The R&D Server Controller console can now be used to disable the servers if someone makes the RD upset.
fix: Fixed the R&D servers working without power.
imageadd: R&D server sprites are now slightly animated, and new sprites have been added for when the server is disabled or off.
/:cl:

image

### CODE REVIEW

**Comment by FloranOtten** [comment]
No longer can a twat somewhere hidden research Biological Engineering, or Cryogenic tech before HE parts. Rejoice!

**Comment by zxaber** [comment]
I personally look forward to the public lynchings of the nanite tech rushers.

**Comment by Tlaltecuhtli** [comment]
what happens if you wear a gasmask with no id

**Comment by zxaber** [comment]
You show up as Unknown; similar to purging an AI with a mask and hidden ID. The console location still shows, though, so whoever checks will know where you were, if not who you were.

File: code/modules/research/rdconsole.dm [hunk 1]
```diff
@@ -167,7 +167,11 @@ Nothing else in the console has ID requirements.
 					var/obj/item/card/id/ID = I.GetID()
 					if(istype(ID))
 						logname = "User: [ID.registered_name]"
-			stored_research.research_logs += "[logname] researched node id [id] with cost [json_encode(price)]."
+			for(var/thing in price)
+				to_chat(world, "DEBUG -- [thing]")
```
Comment (AnturK) [inline code/modules/research/rdconsole.dm]: Ahem

File: code/modules/research/rdconsole.dm [hunk 2]
```diff
@@ -167,7 +167,11 @@ Nothing else in the console has ID requirements.
 					var/obj/item/card/id/ID = I.GetID()
 					if(istype(ID))
 						logname = "User: [ID.registered_name]"
-			stored_research.research_logs += "[logname] researched node id [id] with cost [json_encode(price)]."
+			for(var/thing in price)
+				to_chat(world, "DEBUG -- [thing]")
+			var/i = stored_research.research_logs.len
+			stored_research.research_logs += null
```
Comment (AnturK) [inline code/modules/research/rdconsole.dm]: ???

File: code/modules/research/server.dm [hunk 3]
```diff
@@ -52,14 +71,22 @@
...
 /obj/machinery/rnd/server/proc/mine()
 	. = base_mining_income
 	var/penalty = max((get_env_temp() - temp_tolerance_high), 0) * temp_penalty_coefficient
+	current_temp = get_env_temp()
 	. = max(. - penalty, 0)
 
 /obj/machinery/rnd/server/proc/get_env_temp()
-	var/datum/gas_mixture/environment = loc.return_air()
-	return environment.temperature
+	//var/datum/gas_mixture/environment = loc.return_air()
```
Comment (AnturK) [inline code/modules/research/server.dm]: Why this change ?

**Review by AnturK — CHANGES_REQUESTED** [review]
Debug tchats and some things need explaining.

**Comment by necromanceranne** [comment]
Fucking perfect, we can find out who is being a shit and researching the wrong stuff.

File: code/modules/research/rdconsole.dm (hunk 1 above)
Comment (zxaber) [reply code/modules/research/rdconsole.dm]: Oops, thought I got all these. My bad.

File: code/modules/research/rdconsole.dm (hunk 2 above)
Comment (zxaber) [reply code/modules/research/rdconsole.dm]: I'm adding a whole list as a single element of another list. I can't just use `+=`, because that merges the two lists together. I googled, and found a post by Lummox saying this is how to do it. If there's a better way, I'll gladly change it.

File: code/modules/research/server.dm (hunk 3 above)
Comment (zxaber) [reply code/modules/research/server.dm]: The old code runtimes.

As far as I can tell, many of these procs aren't even used anymore. I suspect that they are relics from before techweb. My guess is that between that and now, atmos code was updated and return_air() no longer returns temperature too.

I change it specifically because I use it to get the temperature for the console readout. In the near future, I want to make servers produce heat again, and slow down if it gets too hot, but I didn't want to hang this PR up on that.

Comment (Astrolight) [reply code/modules/research/server.dm]: Another idea for the heat production would be the ability to overclock or under clock the server for point generation but at the cost of more heat

**Comment by TheSilverNuke** [comment]
You should allow the ability to delete logs as well that way if you have a tator RD or something you can get in there and prevent anyone from finding out it was you that researched illegal tech, hazard nanites, or smth.
